### ISSUE TITLE
Query: inconsistent behavior around anonymous objects comparison

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFrameworkCore#8421]
```
from g in context.Gears
from o in context.Gears.OfType<Officer>()
where new { Name = g.LeaderNickname, Squad = g.LeaderSquadId, Five = 5 } == new { Name = o.Nickname, Squad = o.SquadId, Five = 5 }
select g.Nickname;
```

produces the following SQL:

```sql
SELECT [g].[Nickname]
FROM [Gear] AS [g]
CROSS JOIN [Gear] AS [o]
WHERE [g].[Discriminator] IN (N'Officer', N'Gear') AND ((([g].[LeaderNickname] = [o].[Nickname]) AND ([g].[LeaderSquadId] = [o].[SquadId])) AND (5 = 5))
```
and yields some results, however uncorrelated:

```cs
from g in context.Gears
where new { Five = 5 } == new { Five = 5 }
select g.Nickname
```

produces:

```sql
SELECT [g].[Nickname]
FROM [Gear] AS [g]
WHERE 0 = 1
```

Which always returns false.
This is due to Relinq optimization, generalted QM already has the anonymous type comparison replaced with "false". Linq to objects always returns false when comparing two anonymous objects, however it does seem to be more useful to perform the SQL translation.

If we want to be consistent with L2O then we could collapse anonymous types comparison to ```false```, unless they are part of Join operation - we need to perform proper comparison for joins, e.g. for the composite key join.

**Comment by smitpatel** [comment]
`new { ... } == new { ... }` is always false in L2o world. They are reference equality between 2 different instances (since anonymous types don't implement equals method). Therefore Re-Linq optimizes it. In first query, since the properties are correlated with outer query sources, it was not optimized which eventually got translated to server producing SQL. That is inconsistent. It is possible that if for some reason client eval is forced then we would generate different result.

Comparison of anonymous types behaves differently with `equals` operator for joins. It does property-based comparison for those cases. It is required because you can reference only 1 query source on each side of equals. But for queries like above, it is where predicate. There is no need to introduce anonymous type if the purpose is to do just property comparison since it is easy to write predicate with AND condition. We should detect such cases and either block or in the least case warn just to be more consistent. (or deterministic)

**Comment by divega** [comment]
@maumar @smitpatel @anpete the compiler actually generates an `Equals()` override for each anonymous type to perform property by property comparisons. The `==` operator is not overridden, hence it just performs reference comparisons as @smitpatel mentioned. 

So, I guess the main question in this issue is whether we consider the "normal" semantics of the `==` operator undesirable in queries and therefore we want to compensate. I remember vaguely we have already discussed drifting from the "normal" behavior for `byte[]`, which would be a precedent.

The second (kind of OT) question for me is whether there are any cases in our in-memory implementations in which currently we could be leveraging `Equals()` but we aren't. AFAIK, all standard LINQ to Objects operators that need to perform equality under the hood, e.g. `Join()`, `GroupJoin()`, `GroupBy()` rely on the default `EqualityComparer<T>` which will use the `Equals()` method.

**Comment by divega** [comment]
Clearing up milestone so that we can decide now what we want to do (we believe changing the behavior is a breaking change that we can make in 2.0)

**Comment by maumar** [comment]
Problem is that for q4 (== operator), SqlTranslatingExpressionVisitor performs UnfoldStructuralComparison on binary expressions, which is supposed to convert Expression[] comparison into comparing their constituents. 

In case of Equals method we go thru MethodCallTranslator, which uses EqualsTranslator, which resolves to a Equal BinaryExpression, but UnfoldStructuralComparison is not performed on the result, hence the error.

**Comment by ajcvickers** [comment]
Notes from triage:
* Fix the exception case
* Decide on a consistent behavior

**Comment by smitpatel** [comment]
@maumar Something to investigate again in new pipeline.

**Comment by NetMage** [comment]
Yes, this appears to be a problem again in EF Core 3.1 where it works with `==` on simple value types but fails when encapsulated in anonymous types (and fails when using `Equals` as well), at least in the context of `Where(... Any(... new { } == new { }))`.

**Comment by ajcvickers** [comment]
Discuss with @smitpatel

**Comment by smitpatel** [comment]
> Where(a => Any(b => new { a.fld1 } == new { b.fld1 })) gives a translation error.

It was client eval before 3.0. We haven't regressed anything here.

**Comment by ajcvickers** [comment]
We may consider translating some more anonymous object queries depending on how they are generated--we need to follow up with OData on this.

@NetMage Did you write this manually, or was the LINQ query generated?
